lib/Zabbix/Tiny.pm

```diff
@@ -40,45 +40,73 @@ has 'request'         => ( is => 'ro' );
 has 'json_prepared'   => ( is => 'ro' );
 has 'json_executed'   => ( is => 'ro', default => sub { 0 } );
 has 'redo'            => ( is => 'ro' );
+has 'version'         => ( is => 'rw' );
 
 my @content_type = ( 'content-type', 'application/json', );
 
 sub BUILD {
     my $self      = shift;
     my $ua        = $self->ua;
     my $url       = $self->server;
-    my $json_data = {
-        jsonrpc => '2.0',
-        id      => $self->id,
-        method  => 'user.login',
-        params  => {
-            user     => $self->user,
-            password => $self->password,
-        },
-    };
     if ( $self->verify_hostname == 0 ) {
         $ua->ssl_opts( verify_hostname => 0 );
     }
 
     if ( $self->ssl_opts ) {
         $ua->ssl_opts( %{ $self->{ssl_opts} } );
     }
+
+    if ( !$self->version ) {
+        my $json_data = {
+            jsonrpc => '2.0',
+            id      => $self->id,
+            method  => 'apiinfo.version',
+            params  => {
+            },
+        };
+        my $encoded_json = encode_json ($json_data);
+        my $post_response = $ua->post( $self->server, @content_type,
+        Content => $encoded_json);
+        my $response_content = decode_json( $post_response->{_content} );
+        if ( $response_content->{error} ) {
+            my $error = $response_content->{error}->{data};
+            croak("Error: $error");
+        }
+        else {
+            my @version = split(/\./, $response_content->{'result'});
+            $self->{version} = $version[0];
+        }
+    }
 }
 
 sub login {
     my $self      = shift;
     my $id        = $self->id;
     my $ua        = $self->ua;
     my $url       = $self->server;
-    my $json_data = {
-        jsonrpc => '2.0',
-        id      => $id,
-        method  => 'user.login',
-        params  => {
-            user     => $self->user,
-            password => $self->password,
-        },
-    };
+    my $json_data = "";
+    if ( $self->version < 6 ){
+        $json_data = {
+            jsonrpc => '2.0',
+            id      => $id,
+            method  => 'user.login',
+            params  => {
+                user     => $self->user,
+                password => $self->password,
+            },
+        };
+    } else {
+        $json_data = {
+            jsonrpc => '2.0',
+            id      => $id,
+            method  => 'user.login',
+            params  => {
+                username => $self->user,
+                password => $self->password,
+            },
+        }
+    }
+
```

Comment:
Just reducing some repeition

```suggestion
    my $json_data =  {
        jsonrpc => '2.0',
        id      => $id,
        method  => 'user.login',
    }
        
    if ( $self->version < 6 ) {
        $json_data->{params} = > {
            user     => $self->user,
            password => $self->password,
        };
    } 
    else {
        $json_data->{params} = > {
            username => $self->user,
            password => $self->password,
        };
    }

```

Reply:
Awesome! Adding this change to next commit.